### ISSUE TITLE
Bug: current_url() returning wrong value when different host

### ISSUE THREAD

**Issue by darkylmnx** [codeigniter4/CodeIgniter4#5807]
### PHP Version

8.0

### CodeIgniter4 Version

4.1.9

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

apache

### What happened?

The function `current_url()` returns the `$baseURL` in `app/Config/App.php` instead of the real current url in the browser.
This leads to confusion, know the documentation says:

> You can restrict groups of routes to function only in certain domain or sub-domains of your application by passing the “hostname” option along with the desired domain to allow it on as part of the options array:
> 
> https://codeigniter.com/user_guide/incoming/routing.html?highlight=router#limit-to-hostname

Which means an app could have multiple domain pointing at it, but `current_url()` would only return the one set in config.

Use case: CMS, multi-tenant app, multi-vendor app, etc.

### Steps to Reproduce

Call `current_url()` when loading the app with a host like **sub.example.com** while in the config, you would set `$baseURL` to **www.example.com** for example. 

### Expected Output

`current_url()` is the name implies should retour the **current url** and not the **current url with $baseURL as host**.

**Comment by kenjis** [comment]
> All of the `*_url()` functions are intentionally based on the Config (i.e. baseURL, indexPage, forceSecureRequests) and not intended to match the browser URL. 
https://github.com/codeigniter4/CodeIgniter4/issues/3004#issuecomment-835342299

**Comment by kenjis** [comment]
At least `current_url()` assumes the`$baseURL` is correct.
It makes sense.

But the problem is there is no way to set the currect value to `$baseURL`
when you have more than one hostname.

**Comment by kenjis** [comment]
@darkylmnx 
You need to set the correct `$baseURL`.
But be careful not to create security holes, please.
See
- https://expressionengine.com/blog/http-host-and-server-name-security-issues
- https://www.codeigniter.com/userguide3/installation/upgrade_303.html#step-2-make-sure-your-base-url-config-value-is-not-empty

You could set it in the constructor something like this:

```php
<?php

namespace Config;

use CodeIgniter\Config\BaseConfig;

class App extends BaseConfig
{
    // ...

    public function __construct()
    {
        parent::__construct();

        $allowed_domains = ['domain1.tld', 'domain2.tld'];
        if (in_array($_SERVER['HTTP_HOST'], $allowed_domains, true)) {
             $this->baseURL = 'https://' . $_SERVER['HTTP_HOST'] . '/';
        }
    }
```

**Comment by darkylmnx** [comment]
It would make sens it CodeIgniter was designed to be mono-domained which it isn't anymore (because I guess on version 2 it was back then).

As said above, the current docs says you can constrain by hosts when setting up your routes, so it really confuses me that all *_url function act in the opposite way of what the router gives the ability to do.
https://codeigniter.com/user_guide/incoming/routing.html?highlight=router#limit-to-hostname

I think there should at least be an alternative helper to do that.
As for:

> As you have noted, detection is pretty complicated and not guaranteed 

It seems Laravel & Symfony handles it properly, is there something i'm missing here?

**Comment by kenjis** [comment]
Yes, it is very confusing. 
At least I will fix the documentation about `current_url()`. See #5808

If there is something confusing, tell us. I will fix the docs.

> Laravel & Symfony handles it properly

I don't know they really handle properly. Did you checked the implementation?
And how Symfony handles multiple domains?

**Comment by darkylmnx** [comment]
@kenjis thanks for the docs.
I did check because I use Laravel in another project. Basically it uses Symfony's `Request` class and narrows down to the methods `getPathInfo` & `getHost` :
 https://github.com/symfony/symfony/blob/5e4151fd88260adeac7de0425397b720233c31a2/src/Symfony/Component/HttpFoundation/Request.php#L1681

https://github.com/symfony/symfony/blob/5e4151fd88260adeac7de0425397b720233c31a2/src/Symfony/Component/HttpFoundation/Request.php#L1114

https://github.com/symfony/symfony/blob/5e4151fd88260adeac7de0425397b720233c31a2/src/Symfony/Component/HttpFoundation/Request.php#L854

And finally:

https://github.com/symfony/symfony/blob/5e4151fd88260adeac7de0425397b720233c31a2/src/Symfony/Component/HttpFoundation/Request.php#L1002

In my usage, I never had any problem with Laravel's `url()->current()` or `url()->full()` methods which are based on the above.

From what I read and understand in the Symfony code (as this is way above what I do in PHP), it seems to handle most cases, even edge cases like PHP loaded by CGI and thigns like that.

It's way better than what I did to handle this myself in CI 4 currently, since I'm building a multi-vendor app where each users can connect his/her custom domain, and therefore, creating an allow-list and remplacing `$this->baseURL` like you did in the example doesn't work.

I could indeed check the DB instead of doing an allow-list and if a match is found replace it, but in any case, I would have other types of issues in my specific case.

I actually had to create a method that does this:
```php
public function getCurrentUrl() {
        $protocol = isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? 'https' : 'http';
        $path = "://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";

        return $protocol . $path;
    }
```

But I'd prefer the framework to have a built-in-more-secure-[insert anything]-way of doing this.
I could load Symfony's standalone Request class for that, but I would loose the whole point of using CI4, speed and simplicity.

**Comment by kenjis** [comment]
Thank you for your investigation.

It seems Symfony uses `$trustedHostPatterns` to check the correct hostname.
How do you set the values in Lavarel?

> I could indeed check the DB instead of doing an allow-list and if a match is found replace it, but in any case, I would have other types of issues in my specific case.

What do you mean? If you don't check the hostname, attackers could easily attack your app by host header injection.

**Comment by darkylmnx** [comment]
> It seems Symfony uses `$trustedHostPatterns` to check the correct hostname. How do you set the values in Lavarel?

> What do you mean? If you don't check the hostname, attackers could easily attack your app by host header injection.

I never need to set it since I will then have a middleware or whatsoever that will check if the the host of the current url exists in my DB or some data source.

What I meant was, I would not load all my domains and then do a `in_array` that would be way to bloated in the futur.

**Comment by kenjis** [comment]
> What I meant was, I would not load all my domains and then do a in_array that would be way too bloated in the futur.

Okay, the code using `in_array()` is just a sample.

**Comment by darkylmnx** [comment]
> Okay, the code using `in_array()` is just a sample.
Yes, of course. I was just saying thing was to be able to retreive the current url safely.

**Comment by kenjis** [comment]
> That being said, the fundamental issue was to not have a valid current url I can do my check on.

It seems CI4 does not support multiple domain environment completely,
or there is a design flaw.

`$baseURL` is a predefined static value, and CI4 relies on it.

We cannot set `$baseURL` with using DB data, because the Config\App is created
in very early stage of the framework bootstrap.

What I mean is this is a big change, not just adding a new function.

**Comment by kenjis** [comment]
But after all, it seems to me that the only way is to add a new `current_url()` that does not depend on `$baseURL`, as you say.

>  not have a valid current url I can do my check on.

How do you check? Can you show us the new API?
Or can you send a PR?

**Comment by darkylmnx** [comment]
To be clear, I now know this, and feel free to close this issue or at least change it to a feature request.

What i'm saying is, there's no need for it to be tied to db or whatsoever, just having another method like `real_current_url` or `request_current_url` or something would be nice natively.

It would return the REAL url from which the app was loaded.

Just like the Symfony's class which makes no assumptions on DB or anything, it just returns the url and filters it if you set trusted hosts or anything.

But by default, without any trusted hosts, it returns the url it guesses from the request based on HTTP_HOST, SERVER_NAME or SERVER_ADDR.

It's not much of a code of course but having it natively and done in a more secure way than just:
`$host = $_SERVER['HTTP_HOST'] ?: $_SERVER['SERVER_NAME'];` etc, would be nice.

**Comment by kenjis** [comment]
> But by default, without any trusted hosts, it returns the url it guesses from the request based on HTTP_HOST, SERVER_NAME or SERVER_ADDR.

If a user uses it without any trusted hosts, It is a vulnerability. 
It is not secure without checking `$_SERVER['HTTP_HOST']`/`$_SERVER['SERVER_NAME']`.

> It's not much of a code of course but having it natively and done in a more secure way than just:
$host = $_SERVER['HTTP_HOST'] ?: $_SERVER['SERVER_NAME']; etc, would be nice.

I agree it would be nice if we can write such a function.
But I don't know how to write it in a secure way.

**Comment by darkylmnx** [comment]
@kenjis the vulnerability is only if the method is used by CI4 without the user knowing, which it is not suppose to.
The current url is something people often need IMO, you ofter need the current path, current route ou base_url.
The base_url should of course not behave the same way but the current_url when it is explicitly asked should be different.

At least, that's how it works in Laravel, be default, If you don't touch any setting, Laravel will return the real current route even when your app's url is set to something else, because Laravel never uses the current url directly, only you can load it.
Laravel users your app url, which is not the same is the current url.

So the vulnerability is not a concerne here as it's the user who should use this feature knowingly (like in my case which is to create a multi-vendor app, I would check the current domain upon my DB to make sure I can trust it).

**Comment by kenjis** [comment]
> At least, that's how it works in Laravel, by default If you don't touch any setting, Laravel will return the real current url even when your app's url is set to something else, because Laravel never uses the current url directly, only you can load it.

I guess Laravel users are creating vulnerable apps every day, aren't they?

I don't think it is good to provide vulnerable functions, so at least I think the new current_url() needs trusted hosts.

**Comment by darkylmnx** [comment]
> I guess Laravel users are creating vulnerable apps every day, aren't they?

If they use the function without proper filters yes, but this is not suppose to happen, thats what I'm saying :)
If it happens, it's on the developer not the FW.

**Comment by kenjis** [comment]
How about creating a Controller filter?
- check the hostname, and update `Config\App::$baseURL`

When `$baseURL` is correct, then `current_url()` returns the correct value.

**Comment by MGatner** [comment]
Chiming I'm pretty late here... let's go with @kenjis doc fix for now and handle this properly in version 5. I'm not opposed to a PR for something like `browser_url()` but just know: it is complicated and fraught. Just look at IncomingRequest and the mess there. 

I worked up an entire PR to split out URI functionality but ultimately there is too much already in place than we can't break. I think there are lots of ways we can improve this for version 5, all under the header "Better support for multiple domains".

**Comment by szajens** [comment]
I started writing a library, it will be ready in a few days, I think it will solve the problem

**Comment by szajens** [comment]
> ### PHP Version
> 8.0
> 
> ### CodeIgniter4 Version
> 4.1.9
> 
> ### CodeIgniter4 Installation Method
> Composer (using `codeigniter4/appstarter`)
> 
> ### Which operating systems have you tested for this bug?
> Windows
> 
> ### Which server did you use?
> apache
> 
> ### Database
> _No response_
> 
> ### What happened?
> The function `current_url()` returns the `$baseURL` in `app/Config/App.php` instead of the real current url in the browser. This leads to confusion, know the documentation says:
> 
> > You can restrict groups of routes to function only in certain domain or sub-domains of your application by passing the “hostname” option along with the desired domain to allow it on as part of the options array:
> > https://codeigniter.com/user_guide/incoming/routing.html?highlight=router#limit-to-hostname
> 
> Which means an app could have multiple domain pointing at it, but `current_url()` would only return the one set in config.
> 
> Use case: CMS, multi-tenant app, multi-vendor app, etc.
> 
> ### Steps to Reproduce
> Call `current_url()` when loading the app with a host like **sub.example.com** while in the config, you would set `$baseURL` to **[www.example.com](http://www.example.com)** for example.
> 
> ### Expected Output
> `current_url()` is the name implies should retour the **current url** and not the **current url with $baseURL as host**.
> 
> ### Anything else?
> _No response_

I started writing a library, it will be ready in a few days, I think it will solve the problem

**Comment by MGatner** [comment]
@szajens Please share once you have it done! I'm curious about your solution.

**Comment by szajens** [comment]
> @szajens Please share once you have it done! I'm curious about your solution.

[PECE Routing for CI4](https://github.com/szajens/Pece-for-Codeigniter4)

I'm sorry it took so long, but I had a lot of other work.

I still have to put the library in order.
I have a problem with English, and I will need help from someone who knows English well to correct the README.md

I need to work on the url_to helper in next version

**Comment by MGatner** [comment]
Thanks for sharing! I like the solution. Let me make a free suggestions on the repo structure - mostly, I think that you could release this as its own library that developers could include via Composer, instead of having to merge files into their **app/** directory. 

1. To do this you would need **composer.json**. But this also allows you to specify authorship, set PHP versions, etc
2. Your Config files don't need to copy the App versions, they can have just your content. For example Events only needs your listening. Here is an example of a library Events.php: https://github.com/tattersoftware/codeigniter4-outbox/blob/develop/src/Config/Events.php
3. Tests! Adding tests covers you (and any changes you make in the future), gives confidence to developers who want to use your library, and model how your classes work. Even if they are basic to start getting something in there is a good practice.

**Comment by datamweb** [comment]
Hello @szajens , I am a beginner like you.
I did not use a composer, and I had no knowledge of unit testing. I am learning recently.
When I first used the composer, everything was vague to me. But the more I use it, the more I regret why I did not know it before. I suggest you use it.
To learn the unit test, because I was confused recently, I suggest you follow the steps below in order.
1. Read a short article about it [here](https://andy-carter.com/blog/phpunit-what-why-how)
2. Watch the 10-part unit test video [here](https://www.youtube.com/watch?v=k9ak_rv9X0Y)
3. For more mastery of the official documentation of the unit, see [here](https://phpunit.de/documentation.html).
4. From now on, write a unit test at the same time for all the code you write.
If you have a problem, I will be with you through [forum](https://forum.codeigniter.com/), of course, as far as I know.
